sympy/polys/domains/finitefield.py

```diff
@@ -202,5 +203,21 @@ def from_RealField(K1, a, K0):
         if q == 1:
             return K1.dtype(K1.dom.dtype(p))
 
+    def is_square(self, a):
+        """Returns True if ``a`` is a square modulo p. """
+        # TODO: Faster algorithm without computing sqrt, for example using
+        # Legendre symbol
+        return self.exsqrt(a) is not None
+
+    def exsqrt(self, a):
+        """Returns the square root of ``a`` modulo p that is no larger than
+        ``p // 2``, or ``None`` if it does not exist. """
+        # Factorize x**2 - a
+        poly = [x.val for x in [self.one, self.zero, -a]]
+        for factor in gf_zassenhaus(poly, self.mod, self.dom):
```

Comment:
This approach can also be used for `AlgebraicField` with `dup_ext_factor`.

Reply:
Following the comment by @asmeurer, we should call the lowest level routine possible. I see that the `dup_ext_factor` is still doing a lot of checks and preparations, so I figure we should use another more lower level routine for our special case (where the polynomial is monic and of degree 2).

For example, for finite field factorization, I call the lower level `gf_zassenhaus` instead of the more obvious `gf_factor` because I know the polynomial is monic and square-free (unless a=0 or the field is characteristic 2, which I handled separately).

But I can't understand the `dup_ext_factor` algorithm probably because I'm not well-versed in algebraic number theory. In particular I have no idea what `dup_sqf_norm` is doing.

Do you think it's possible to use some lower level routines than `dup_ext_factor` for our special case?